design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,243 @@
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+spec:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+  provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+4. An admin creates a new `MySQLInstancePolicy` object, specifying that the previously created resource class (`kind: RDSInstanceClass`) should serve as default for a given claim kind `mysqlinstance.storage.crossplane.io`. A `MySQLInstancePolicy` could look as follows:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: MySQLInstancePolicy
+metadata:
+  name: mysql-aws-policy
+  namespace: demo
+defaultClassRef:
+  kind: rdsinstanceclass.database.aws.crossplane.io/v1alpha1
+  name: standard-mysql
+  namespace: crossplane-system
+```
+
+5. A developer creates a new resource claim (`kind: MySQLInstance`), omitting a class reference so as to fall back on the class defined as default for the claim kind:
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  engineVersion: "9.6"
+```
+
+6. Upon creation, the `MySQLInstance` default class controller searches for `MySQLInstancePolicy` and `MySQLInstanceClusterPolicy` objects. In this case, there is a single `MySQLInstancePolicy` that exists in the same namespace as the `MySQLInstance` claim, so the claim's `classRef` is set to match the `defaultClassref` of the `MySQLInstancePolicy`. The following scenarios could have occurred to cause a different outcome:
+    * If there had been multiple `MySQLInstancePolicy` instances in the claim's namespace, the claim would have failed to be reconciled by the policy controller because it would not be able to choose only one, despite the presence of a `MySQLInstanceClusterPolicy` or not (i.e. an undecidable namespace level policy situation does not fall back on a decidable cluster level policy, it just fails).
+    * If there had been no instances of `MySQLInstancePolicy` in the claim's namespace, but there had been a `MySQLInstanceClusterPolicy` instance defined in the claim's cluster, the claim's `classRef` would have been set to the `MySQLInstanceClusterPolicy` instance's `defaultClassRef`.
+    * If there had been no instances of `MySQLInstancePolicy` in the claim's namespace, but there had been multiple instances of `MySQLInstanceClusterPolicy` defined in the claim's cluster, the claim would have failed to be reconciled by the policy controller because it would not be able to choose only one.
+7. The claim is reconciled by the claim controller now that it passes the class reference predicate.
```

Comment:
I think there would be a little value in showing an example of a resource claim referencing a strongly typed resource class, which would now be a `*corev1.ObjectReference` with its `APIVersion` and `Kind` fields populated.

This is a nice property for the resource claim reconciler, which can switch from using the `resource.ObjectHasProvisioner` predicate (which must look up the claim's class in order to determine its provisioner) to a hypothetical `resource.ObjectHasClassKind` predicate.

Reply:
Agreed. Also, we will want to make sure to update this document if / when resource classes become cluster-scoped.